### PR TITLE
Security features

### PR DESCRIPTION
**Pull request by paddeo**

### CODE REVIEW

**Comment by entando-service** [comment]
Something went wrong. Investigate!

**Comment by entando-service** [comment]
All is well

**Comment by entando-service** [comment]
All is well

**Comment by entando-service** [comment]
Something went wrong. Investigate!

**Comment by entando-service** [comment]
Something went wrong. Investigate!

File: engine/src/main/java/com/agiletec/aps/system/services/baseconfig/BaseConfigManager.java [hunk 1]
```diff
@@ -37,92 +48,337 @@
...
-	}
+    private static final Logger _logger = LoggerFactory.getLogger(BaseConfigManager.class);
+
+    @Override
+    public void init() throws Exception {
+        String version = this.getSystemParams().get(SystemConstants.INIT_PROP_CONFIG_VERSION);
+        this.getCacheWrapper().initCache(this.getConfigDAO(), version);
+        _argon2 = (this.getParam("argon2") != null
+                && this.getParam("argon2").equalsIgnoreCase("true"));
+        _logger.debug("{} ready. Initialized", this.getClass().getName());
```
[the remark this answers is not in the document]
Comment (eugeniosant) [reply engine/src/main/java/com/agiletec/aps/system/services/baseconfig/BaseConfigManager.java]: fixed

File: engine/src/main/java/com/agiletec/aps/system/services/baseconfig/BaseConfigManager.java [hunk 2]
```diff
@@ -37,92 +48,337 @@
...
+     * Map contenente tutte le voci di configurazione di una versione.
+     */
+    private Map<String, String> _systemParams;
+
+    /**
+     * Map contenente tutti i parametri di configurazione di una versione.
+     */
+    private IConfigManagerCacheWrapper _cacheWrapper;
+
+    private String _keyString;
```
[the remark this answers is not in the document]
Comment (eugeniosant) [reply engine/src/main/java/com/agiletec/aps/system/services/baseconfig/BaseConfigManager.java]: Parameters removed from the service

File: engine/src/main/java/com/agiletec/aps/system/services/baseconfig/BaseConfigManager.java [hunk 3]
```diff
@@ -37,92 +48,337 @@
...
+
+            // Try loading properties from the file (if found)
+            props.load(is);
+        } catch (Exception e) {
+            _logger.error("Error getting security.properties");
+        }
+        return props;
+    }
+
+    public void saveParamChanges() throws Exception {
```
[the remark this answers is not in the document]
Comment (eugeniosant) [reply engine/src/main/java/com/agiletec/aps/system/services/baseconfig/BaseConfigManager.java]: Fixed

File: engine/src/main/java/com/agiletec/aps/system/services/baseconfig/BaseConfigManager.java [hunk 4]
```diff
@@ -37,92 +48,337 @@
...
+        this.setKeyString(Base64.getEncoder().encodeToString(secretKey.getEncoded()));
+        System.getProperties().setProperty("key.string.encryption", this.getKeyString());
+        try {
+            Properties props = new Properties();
+            props.setProperty("key.string.encryption", this.getKeyString());
+            File f = new File(this.getConfigFolder() + File.separator + "security.properties");
+            OutputStream out = new FileOutputStream(f);
+            props.store(out, "DO NOT EDIT THIS FILE");
+        } catch (Exception e) {
+            e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (eugeniosant) [reply engine/src/main/java/com/agiletec/aps/system/services/baseconfig/BaseConfigManager.java]: Fixed

File: engine/src/main/java/com/agiletec/aps/system/services/baseconfig/BaseConfigManager.java [hunk 5]
```diff
@@ -37,92 +48,337 @@
...
+    public String getParam(String name) {
+        String param = this.getSystemParams().get(name);
+        if (null != param) {
+            return param;
+        } else {
+            return this.getCacheWrapper().getParam(name);
+        }
+    }
+
+    public void keyGenerate() throws Exception {
```
[the remark this answers is not in the document]
Comment (eugeniosant) [reply engine/src/main/java/com/agiletec/aps/system/services/baseconfig/BaseConfigManager.java]: Deleted!

**Comment by entando-service** [comment]
Something went wrong. Investigate!

**Comment by entando-service** [comment]
Something went wrong. Investigate!

**Comment by entando-service** [comment]
All is well

**Comment by entando-service** [comment]
All is well

File: engine/src/main/java/com/agiletec/aps/system/services/baseconfig/BaseConfigManager.java [hunk 6]
```diff
@@ -37,92 +48,337 @@
...
+            props.setProperty("algo.argon2.hash.length", String.valueOf(this.getHashLen()));
+            props.setProperty("algo.argon2.salt.length", String.valueOf(this.getSaltLen()));
+            props.setProperty("algo.argon2.iterations", String.valueOf(this.getIterations()));
+            props.setProperty("algo.argon2.memory", String.valueOf(this.getMemory()));
+            props.setProperty("algo.argon2.parallelism", String.valueOf(this.getParallelism()));
+            File f = new File(this.getConfigFolder() + File.separator + "security.properties");
+            OutputStream out = new FileOutputStream(f);
+            props.store(out, "DO NOT EDIT THIS FILE");
+        } catch (Exception e) {
+            e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (eugeniosant) [reply engine/src/main/java/com/agiletec/aps/system/services/baseconfig/BaseConfigManager.java]: Removed method

File: engine/src/main/db/update/scriptServ.sql [hunk 7]
```diff
@@ -55,5 +55,9 @@ CREATE TABLE dataobjectmodels
   CONSTRAINT dataobjectmodels_pkey PRIMARY KEY (modelid )
 );
 
-ALTER TABLE actionlogcommentrecords RENAME comment TO commenttext;
```
[the remark this answers is not in the document]
Comment (eugeniosant) [reply engine/src/main/db/update/scriptServ.sql]: Restored script

**Comment by entando-service** [comment]
All is well

File: engine/src/main/java/com/agiletec/aps/system/services/user/UserManager.java [hunk 8]
```diff
@@ -22,229 +22,287 @@
...
-			}
-		} catch (Throwable t) {
-			_logger.error("Error searching users by text '{}'", text,  t);
-			throw new ApsSystemException("Error loading the user list", t);
-		}
-		return users;
-	}
+    @Override
+    public void init() throws Exception {
+        User admin = (User) this.getUser("admin");
```
[the remark this answers is not in the document]
Comment (paddeo) [reply engine/src/main/java/com/agiletec/aps/system/services/user/UserManager.java]: I changed it.

File: engine/src/main/java/com/agiletec/aps/system/services/user/UserManager.java [hunk 9]
```diff
@@ -22,229 +22,287 @@
...
-				this.setUserCredentialCheckParams(users.get(i));
-			}
-		} catch (Throwable t) {
-			_logger.error("Error searching users by text '{}'", text,  t);
-			throw new ApsSystemException("Error loading the user list", t);
-		}
-		return users;
-	}
+    @Override
+    public void init() throws Exception {
```
[the remark this answers is not in the document]
Comment (paddeo) [reply engine/src/main/java/com/agiletec/aps/system/services/user/UserManager.java]: I changed the method. Now at every start of the application the method checks the admin user password encryption (if admin user exists). Then, if the system is not already on Argon2 encryption, the method change password encryption for users that do not already have Argon2.

File: engine/src/main/java/com/agiletec/aps/system/services/user/UserManager.java [hunk 10]
```diff
@@ -22,229 +22,287 @@
...
+        User admin = (User) this.getUser("admin");
+        if (admin.isDisabled()) {
```
[the remark this answers is not in the document]
Comment (paddeo) [reply engine/src/main/java/com/agiletec/aps/system/services/user/UserManager.java]: I added a check.

File: engine/src/main/java/com/agiletec/aps/system/services/user/UserManager.java [hunk 11]
```diff
@@ -22,229 +22,287 @@
...
+            admin.setDisabled(false);
+            this.getUserDAO().updateUser(admin);
+        }
+        if (this.getUserDAO().getEncrypter() instanceof Argon2Encrypter
+                && !this.getConfigManager().isArgon2()) {
+            List<UserDetails> users = this.getUserDAO().loadUsers();
+            for (UserDetails user : users) {
+                User cur = (User) user;
+                String pwd = cur.getPassword();
+                if (!pwd.contains("argon2")) {
```
[the remark this answers is not in the document]
Comment (paddeo) [reply engine/src/main/java/com/agiletec/aps/system/services/user/UserManager.java]: I fixed that check.

**Comment by entando-service** [comment]
All is well

**Pull request merged**